src/PathFinding.h

```diff
@@ -0,0 +1,91 @@
+#include <vector>
+#include <map>
+#include <functional>
+#include <algorithm>
+#include <limits>
+
+#ifndef PATH_FINDING_H
+#define PATH_FINDING_H
+
+template<typename Node, typename Key>
+std::vector<Node*> find_shortest_path(
+    Node& start,
+    std::function<Key (Node&)> get_key,
```

Comment:
I'm not used to seeing this many `std::function` used as delegates, but it's possible it's due to the kind of C++ code I usually look at.

If you want to play with C++ 20 concepts, you could have a `PathfindingQuery` concept for a struct with those three methods and typedefs/usings for the node and key types. That'd be cool.